keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,544 @@
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the asthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+asthetics/egronomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, deployment *appsv1apply.Deployment, fieldManager string, metav1.ApplyOptions) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
+  DryRun []string `json:"dryRun,omitempty" protobuf:"bytes,1,rep,name=dryRun"`
+  Force *bool `json:"force,omitempty" protobuf:"varint,2,opt,name=force"`
+}
+
+func (ApplyOptions) ToPatchOptions(fieldManager string) PatchOptions
+```
+
+Apply will combine the `fieldManager` argument with `ApplyOptions` to create the
+`PatchOptions`.
+
+Each apply call will be required to provide a fieldmanager name. We will not
+provide a a way for the fieldmanager name to be set for the entire
+clientset. There are a couple reasons for this:
+
+- If a client has multiple code paths where it makes apply requests to the same
+  object, but with different field sets, they must use different field manager
+  names. If they use the same field manager name they will cause fields to be
+  accidentally removed or disowned. This is a potential foot gun we would like to
+  avoid.
+
+- Apply requests always conflict with update requests, even if they were made by
+  the same client with the same field manager name. This is by design. So when a
+  controller migrates from update to apply, it will need to deal with conflicts
+  regardless of what field manager name is used.
+
+### Generated apply configuration types
+
+All fields present in an apply configuration become owned by the applier after
+when the apply request succeeds. Go structs contain zero valued fields which are
+included even if the user never explicitly sets the field. Required boolean
+fields are a good example of fields that would be applied incorrectly using go
+structs, e.g. `ContainerStatus.Ready` (required, not omitempty). Because of this
```

Comment:
I'm curious how much of a problem this is in practice. If I'm the first to apply an object, I have to specify the required fields anyways. If I'm applying on an existing object, I don't have to apply required fields, but are there scenarios, for example in the `containerStatus.Ready` example here where I would want to apply the `containerStatus` without knowing the ready state? I'd love to see a very convincing example in this doc!

Reply:
We got direct feedback about this in practice when we discussed this with the controller-runtime WG. One interesting thing was that multiple people chimed in reporting that they hadn't realized that there was a problem using the structs with apply until they got burned when a required boolean field got flipped to false during an apply that they had not intended to modify. Interestingly, it was specifically the boolean typed fields that they reported.

The `ContainerStatus.Ready` example makes a lot of sense to me. If it is already set to true, and you don't intend to set it to false when calling apply, it is really surprising to find that it was set to its default of false, which is exactly what can happen if using the existing go structs.

Reply:
Per our offline discussion I agree `ContainerStatus.Ready` is not an ideal example since it is difficult to imagine a usage pattern when it would not be included in the apply request.

I've rewritten this section to include more examples. I think `ContainerPort.ContainerPort` is a slightly better example, but anyone can review a larger list using a grep like:

```sh
find . -name "types.go" | xargs git grep -E " (int[0-9]{0,2}|bool) \`json:" | grep -v omitempty |grep -v vendor | grep -v "*"
```

I also did some searching and found that we have well over 100 fields like this, which I think adds weight to there being a fundamental soundness problem.